Client/game_sa/CWeaponSA.cpp

```diff
@@ -20,230 +19,83 @@
 
 extern CGameSA* pGame;
 
-CWeaponSA::CWeaponSA(CWeaponSAInterface* pInterface, CPed* pOwner, eWeaponSlot weaponSlot)
-    : m_pInterface{pInterface}, m_pOwner{pOwner}, m_weaponSlot{weaponSlot}
+CWeaponInfo* CWeaponSA::GetInfo(const eWeaponSkill& skill) const
 {
-}
-
-void CWeaponSA::Destroy()
-{
-    if (!m_pOwner)
-    {
-        delete m_pInterface;
-        delete this;
-    }
-}
-
-eWeaponType CWeaponSA::GetType()
-{
-    return m_pInterface->m_eWeaponType;
-};
-
-void CWeaponSA::SetType(eWeaponType type)
-{
-    m_pInterface->m_eWeaponType = type;
-}
-
-eWeaponState CWeaponSA::GetState()
-{
-    return m_pInterface->m_eState;
-}
-
-void CWeaponSA::SetState(eWeaponState state)
-{
-    m_pInterface->m_eState = state;
-}
-
-DWORD CWeaponSA::GetAmmoInClip()
-{
-    return m_pInterface->m_nAmmoInClip;
-}
-
-void CWeaponSA::SetAmmoInClip(DWORD dwAmmoInClip)
-{
-    m_pInterface->m_nAmmoInClip = dwAmmoInClip;
-}
-
-DWORD CWeaponSA::GetAmmoTotal()
-{
-    return m_pInterface->m_nAmmoTotal;
-}
-
-void CWeaponSA::SetAmmoTotal(DWORD dwAmmoTotal)
-{
-    m_pInterface->m_nAmmoTotal = dwAmmoTotal;
-}
-
-CPed* CWeaponSA::GetPed()
-{
-    return m_pOwner;
-}
-
-eWeaponSlot CWeaponSA::GetSlot()
-{
-    return m_weaponSlot;
+    return m_interface ? pGame->GetWeaponInfo(m_interface->m_eWeaponType, skill) : nullptr;
 }
 
 void CWeaponSA::SetAsCurrentWeapon()
 {
-    m_pOwner->SetCurrentWeaponSlot(m_weaponSlot);
+    if (m_owner)
+        m_owner->SetCurrentWeaponSlot(m_weaponSlot);
 }
 
-CWeaponInfo* CWeaponSA::GetInfo(eWeaponSkill skill)
+void CWeaponSA::Destroy()
 {
-    return pGame->GetWeaponInfo(m_pInterface->m_eWeaponType, skill);
+    if (m_interface)
+        delete m_interface;
+
+    delete this;
 }
 
 void CWeaponSA::Remove()
 {
-    DWORD dwFunc = FUNC_Shutdown;
-    DWORD dwThis = (DWORD)m_pInterface;
-    _asm
-    {
-        mov     ecx, dwThis
-        call    dwFunc
-    }
+    if (!m_interface)
+        return;
+
+    m_interface->Shutdown();
 
     // If the removed weapon was the currently active weapon, switch to empty-handed
-    if (m_pOwner->GetCurrentWeaponSlot() == m_weaponSlot)
+    if (m_owner && m_owner->GetCurrentWeaponSlot() == m_weaponSlot)
```

Comment:
Is is possible that the owner is nullptr?

Reply:
I've discussed the situation around "Is it possible that (x) is null?" cases with him on Discord, there are particular security reasons for additional checks that may seem redundant.

Reply:
These check are useless and do nothing